### PR TITLE
[FIXED JENKINS-31871] Properly handle single quotes in item names

### PR DESCRIPTION
**Pull request by fbelzunc**
Single quotes in item names should be prohibited
https://issues.jenkins-ci.org/browse/JENKINS-31871

@reviewbybees @oleg-nenashev @varmenise

### CODE REVIEW

File: core/src/main/java/jenkins/model/ProjectNamingStrategy.java [hunk 1]
```diff
@@ -99,6 +101,11 @@ public DefaultProjectNamingStrategy() {
         @Override
         public void checkName(String origName) throws Failure {
             // default - should just do nothing (this is how Jenkins worked before introducing this ExtensionPoint)
+            if (StringUtils.isNotBlank(DEFAULT_PATTERN) && StringUtils.isNotBlank(origName)) {
```
Comment (KostyaSha) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java]: why not just check string contains?

File: core/src/main/resources/jenkins/model/Messages.properties [hunk 2]
```diff
@@ -34,6 +34,7 @@ Hudson.NoName=No name is specified
 Hudson.NodeBeingRemoved=Node is being removed
 Hudson.UnsafeChar=\u2018{0}\u2019 is an unsafe character
 Hudson.JobNameConventionNotApplyed=\u2018{0}\u2019 does not match the job name convention pattern {1}
+Hudson.JobSinglesQuotesAreProhibited=Singles quotes are prohibited
```
Comment (oleg-nenashev) [inline core/src/main/resources/jenkins/model/Messages.properties]: Needs a better explanation

File: core/src/main/java/jenkins/model/ProjectNamingStrategy.java (hunk 1 above)
Comment (fbelzunc) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java]: Yes, this is true ;-) Copy/Paste issue xD

**Comment by oleg-nenashev** [comment]
This approach may be a valid workaround for the issue. It could be also extended to other such symbols later. I would also recommend to introduce the verification in <code> PatternProjectNamingStrategy</code> as well.

File: core/src/main/java/jenkins/model/ProjectNamingStrategy.java (hunk 1 above)
Comment (fbelzunc) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java]: I think I am going just to do:

```
@Override
        public void checkName(String origName) throws Failure {
            // default - should just do nothing (this is how Jenkins worked before introducing this ExtensionPoint)
            if (StringUtils.isNotBlank(origName) && !Pattern.matches(DEFAULT_PATTERN, origName)) {
                throw new Failure(Messages.Hudson_JobSinglesQuotesAreProhibited());
            }
        }
```

> why not just check string contains?

I am following the way the pattern is used on the rest of the methods. I think this should be fine.

Comment (KostyaSha) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java]: imho, check for blank string -> throw blank error, check for single quote -> throw single quote. As job tied to FS probably this should be done with some additional restrictions like `/`

Comment (KostyaSha) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java]: and afair this strategy can be changed in global config, shouldn't strict check be done in some other routine i.e. `ProjectNamingStrategy.checkName() and call from implementations to super.checkname()`?

Comment (KostyaSha) [inline core/src/main/java/jenkins/model/ProjectNamingStrategy.java]: > I am following the way the pattern is used on the rest of the methods. I think this should be fine.

Pattern is time consuming instance. Use it when it really required.

>I think I am going just to do:
Enter space -> get single quote error --> not logical for end users See my comment https://github.com/jenkinsci/jenkins/pull/1943/files#r46538967

**Comment by fbelzunc** [comment]
@oleg-nenashev @KostyaSha What do you think now?

**Comment by KostyaSha** [comment]
Sorry if i missed something, but I see no description about root of issue with single quotes, is it FS limitations? Does jenkins has any mechanism for project naming restriction contributed from FS (some item*storage* in future)? In case of FS limitation changes to default wouldn't make sense because it impossible to have multiple restrictions (I still didn't get answer https://groups.google.com/d/msg/jenkinsci-dev/KvOX6v1DKUY/rIDsYQHS8SYJ) and such checks must be enforced in other places imho.

**Comment by KostyaSha** [comment]
@fbelzunc as usual imho core should provide good APIs and logic instead of "hacks" like allow create things that you can't access and work with later. That's a bug imho that should be fixed in the right way (some Item name check Extension (storage limitation? web path limitation?) with checks before doing project naming limitations i.e.) .

**Comment by fbelzunc** [comment]
@KostyaSha 

> Sorry if i missed something, but I see no description about root of issue with single quotes, is it FS limitations?

I think this happens in all Unix OS - AFAIK you cannot create files/directories with single quotes.

> Does jenkins has any mechanism for project naming restriction contributed from FS (some item*storage* in future)?

I have no idea about this.

> In case of FS limitation changes to default wouldn't make sense because it impossible to have multiple restrictions

I agree but at least I think that to use singles quotes is a bad practice and error prone.

> as usual imho core should provide good APIs and logic instead of "hacks" like allow create things that you can't access and work with later. That's a bug imho that should be fixed in the right way (some Item name check Extension (storage limitation? web path limitation?) with checks before doing project naming limitations i.e.) .

I am not sure what to say here. I think what you say makes sense, but I would let the rest to give their opinion. For me this change should be fine, I understand you consider it as a hack, and I guess you might prefer an Extension point checking the FS in order not to use a default patterns to avoid. I think we should just try to avoid crazy characters like [these ones](http://www.mtu.edu/umc/services/web/cms/characters-avoid/).

**Comment by KostyaSha** [comment]
> checking the FS

As soon as many people wants non FS backend (cc @abayer ) core should provide more generic extension i.e. some new `ItemNameRestriction` or `ProjectNamingStrategy` with ability to have order/be invisible (i.e. web path(url) restriction and backend naming rules should be checked before any end user preferences). I don't think that it will take too much time to resolve this part in core in right way.

**Comment by fbelzunc** [comment]
I was wrong it might no be related to the FS, I was able to use `touch test\'file.txt` and create the file with the single quotes. So the root cause needs to be investigated.

File: core/src/main/resources/jenkins/model/Messages.properties [hunk 3]
```diff
@@ -34,6 +34,7 @@ Hudson.NoName=No name is specified
 Hudson.NodeBeingRemoved=Node is being removed
 Hudson.UnsafeChar=\u2018{0}\u2019 is an unsafe character
 Hudson.JobNameConventionNotApplyed=\u2018{0}\u2019 does not match the job name convention pattern {1}
+Hudson.JobSinglesQuotesAreProhibited=Singles quotes are prohibited. You won't be able to reliably access it.
```
Comment (jglick) [inline core/src/main/resources/jenkins/model/Messages.properties]: `s/'/’/` :-)

**Comment by jglick** [comment]
This is not a fix for the reported issue. The fix would be to correctly URL-encode all characters, no matter how foolhardy.

And modifying the default project naming strategy is incompatible. It is lenient for historical reasons. If you want to guide people to use a particular recommended strategy, there are various options:

* add an `AdministrativeMonitor` when no strategy is set
* set one on a new instance out of the box
* add form validation for the _New Item_ field _Name_ if a dangerous-seeming character is in use but the current strategy permits it

**Comment by abayer** [comment]
I agree with @jglick - regardless of the backend storage (yup, @KostyaSha, I'm still dreaming there!), URL-encoding is the right thing to do for canonical representations of job names.

**Comment by daniel-beck** [comment]
:-1: This does not solve the reported problem (which I cannot even reproduce entirely, so I doubt it's only a problem in core) but is a terrible workaround.

**Comment by fbelzunc** [comment]
@daniel-beck @jglick @KostyaSha The main issue here is with the Form validation. 

Removing `jsStringEscape` in `FormValidations#toStemUrl` do the trick.

I checked that:

1. With singles quotes FormValidation works
2. With double-quotes  FormValidation works
3. Double // is already un unsafe character by [this line] (https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/model/Jenkins.java#L3137) - so not possible to use it.

However, I don't fully understand why when I don't scape those characters then everything works. 

* Any ideas?

**Comment by oleg-nenashev** [comment]
I would say it happens due to the double escaping. Firstly you escape the symbol for JS, and this step legitimately adds the slash. Then the framework method processes the URL and sees the valid characters (/') => it does not escape anything else. But we get the undesired slash

**Comment by daniel-beck** [comment]
I'm pretty sure the reason for this is in the weird `checkUrl` syntax -- it is actually supposed to be valid JavaScript in `hudson-behavior.js`'s `registerValidator`. With this change, does it complain about the legacy URL not being valid JS?

Alternatively, we could just clean this up. It's weird.

**Comment by fbelzunc** [comment]
> With this change, does it complain about the legacy URL not being valid JS?

For me the change works fine, it doesn't complaint about anything.

**Comment by daniel-beck** [comment]
Given how trivial the final fix is, should be rebased into a single commit with proper commit message referencing the issue it fixes before merging.

**Comment by daniel-beck** [comment]
In my limited testing, this worked well so :+1: from me. I wonder what happens to the 'legacy' behavior implemented in `hudson-behavior.js` though…

**Comment by fbelzunc** [comment]
@daniel-beck You have now only one commit.

**Comment by daniel-beck** [comment]
@jenkinsci/code-reviewers Any of you know how to [get into this legacy behavior](https://github.com/jenkinsci/jenkins/blob/f71725202159a7fd1ba588c30ee0602477533370/war/src/main/webapp/scripts/hudson-behavior.js#L423)?

**Comment by oleg-nenashev** [comment]
@daniel-beck Nope. Seems to be a dead code after quick check

**Comment by daniel-beck** [comment]
@oleg-nenashev Soooo… merge and see what happens? Or WDYT?

**Comment by KostyaSha** [comment]
> @jenkinsci/code-reviewers Any of you know how to get into this legacy behavior?

https://github.com/jenkinsci/jenkins/commit/afa517e4615371d9a151ef2867dd385093fc8619

**Comment by daniel-beck** [comment]
@oleg-nenashev (and @jenkinsci/code-reviewers ) Ping

**Comment by oleg-nenashev** [comment]
@daniel-beck So @KostyaSha provided a link to the issue, which contains some repro steps. Maybe needs a follow-up

I'm fine with the merge

**Review by stephenc — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@daniel-beck I'm fine with merge

**Comment by oleg-nenashev** [comment]
2 votes for the merge, I'm ambivalent.
Merging according to the voting results.

**Pull request merged**